hw/opentitan/ot_hmac.c

```diff
@@ -259,6 +277,102 @@ struct OtHMACState {
     char *ot_id;
 };
 
+static inline enum OtHMACDigestSize ot_hmac_get_digest_size(uint32_t cfg_reg)
+{
+    switch ((cfg_reg & R_CFG_DIGEST_SIZE_MASK) >> R_CFG_DIGEST_SIZE_SHIFT) {
+    case 0x1u:
+        return HMAC_SHA2_256;
+    case 0x2u:
+        return HMAC_SHA2_384;
+    case 0x4u:
+        return HMAC_SHA2_512;
+    case 0x8u:
+    default:
+        return HMAC_SHA2_NONE;
+    }
+}
+
+static inline size_t ot_hmac_get_digest_bytes(enum OtHMACDigestSize digest_size)
+{
+    switch (digest_size) {
+    case HMAC_SHA2_256:
+        return 32u;
+    case HMAC_SHA2_384:
+        return 48u;
+    case HMAC_SHA2_512:
+        return 64u;
+    case HMAC_SHA2_NONE:
+    default:
+        qemu_log_mask(LOG_GUEST_ERROR,
+                      "%s: Cannot get digest size if it is not set. \n",
+                      __func__);
+        return 0u;
+    }
+}
+
+static inline size_t ot_hmac_get_block_size_bytes(OtHMACState *s)
+{
+    switch (ot_hmac_get_digest_size(s->regs->cfg)) {
+    case HMAC_SHA2_256:
+        return 64u;
+    case HMAC_SHA2_384:
+    case HMAC_SHA2_512:
+        return 128u;
+    case HMAC_SHA2_NONE:
+    default:
+        qemu_log_mask(LOG_GUEST_ERROR,
+                      "%s: Cannot get block size if digest size is not set. \n",
+                      __func__);
+        return 0u;
+    }
+}
+
+static inline enum OtHMACKeyLength ot_hmac_get_key_length(uint32_t cfg_reg)
+{
+    switch ((cfg_reg & R_CFG_KEY_LENGTH_MASK) >> R_CFG_KEY_LENGTH_SHIFT) {
+    case 0x01u:
+        return HMAC_KEY_128;
+    case 0x02u:
+        return HMAC_KEY_256;
+    case 0x04u:
+        return HMAC_KEY_384;
+    case 0x08u:
+        return HMAC_KEY_512;
+    case 0x10u:
+        return HMAC_KEY_1024;
+    case 0x20u:
+    default:
+        return HMAC_KEY_NONE;
+    }
+}
+
+static inline size_t ot_hmac_get_key_bytes(OtHMACState *s)
+{
+    switch (ot_hmac_get_key_length(s->regs->cfg)) {
```

Comment:
It seems to be a dual dispatch (switch, switch).  As it looks the HW values are always 2^N, so maybe we could use some builtin (I think there are macros in QEMU that map them) to check that only 1 bit is set (`__builtin_popcount`), and shift the valid values to generate the actual keylength / byte count?

(maybe I'm wrong and the HW values do not allow this kind of implementation, please let me know)

Reply:
The HW should allow for this kind of implementation, as the encoding is one-hot. I'll have a look at if I can do this to avoid the dual dispatch.

Reply:
Never mind, I completely missed that this is unfortunately not the case - `HMAC_KEY_384` is `48u` bytes, which is not 2^N. If not for that case then I was going to try something like the below, but this is not possible to optimise away (at least using this method):

```c
static size_t ot_hmac_get_key_bytes(OtHMACState *s)
{
    uint32_t key_length = (cfg_reg & R_CFG_KEY_LENGTH_MASK);
    if (ctpop32(key_length) != 1) {
        g_assert_not_reached();
        return 0u;
    }
    static_assert(R_CFG_KEY_LENGTH_SHIFT > 4);
    return key_length >> (R_CFG_KEY_LENGTH_SHIFT - 4);
}
```

Reply:
Ok, I missed it as well.

Reply:
The other alternative is that, for the three functions with dual dispatch (`ot_hmac_get_digest_bytes`, `ot_hmac_get_block_size_bytes` and `ot_hmac_get_key_bytes`), they all have the same cases as their respective `ot_hmac_get_digest_size` and `ot_hmac_get_key_length` functions. So, I could just eliminate the calls and use the values directly, e.g.:

```c
static size_t ot_hmac_get_key_bytes(OtHMACState *s)
{
    switch ((s->regs->cfg & R_CFG_KEY_LENGTH_MASK) >> R_CFG_KEY_LENGTH_SHIFT) {
    case 0x01u:
        return 16u;
    case 0x02u:
        return 32u;
    case 0x04u:
        return 48u;
    case 0x08u:
        return 64u;
    case 0x10u:
        return 128u;
    case 0x02u:
    default:
        g_assert_not_reached();
        return 0u;
    }
}
```
The benefit to this is the micro-optimisation, which may be worth it if you think it is performance critical. The downside is you lose the single sources of truth for the digest size and key length - now if the hardware or implementation changes, you need to update these values in several places. We could replace this by assigning the enums literal values which we then use in the cases. How do you feel about this approach?

Reply:
Note about `static_assert`: I think clang-tidy will complain that a message is required as message-less assertion is C++ only, _e.g._ 
```c
static_assert(cond, "reason for failure");
```